compiler/rustc_lint/src/builtin.rs

```diff
@@ -2626,16 +2627,17 @@ impl ClashingExternDeclarations {
     fn insert(&mut self, tcx: TyCtxt<'_>, fi: &hir::ForeignItem<'_>) -> Option<HirId> {
         let hid = fi.hir_id;
 
-        let name =
-            &tcx.codegen_fn_attrs(tcx.hir().local_def_id(hid)).link_name.unwrap_or(fi.ident.name);
-
-        if self.seen_decls.contains_key(name) {
+        let local_did = tcx.hir().local_def_id(fi.hir_id);
+        let did = local_did.to_def_id();
+        let instance = Instance::new(did, ty::List::identity_for_item(tcx, did));
+        let name = tcx.symbol_name(instance).name.to_string();
```

Comment:
Could avoid allocating the string here for the `contains_key` and `get` calls.

The if condition below is preexisting but could also be written as:

```rust
if let Some(val) = self.seen_decls.get(name) {
    Some(val)
} else {
    // insert/allocate here.
}
```

It could also use the `HashMap::entry` API to avoid the double-lookup (though it would need to allocate the string unconditionally I believe).